### PR TITLE
feat(vanilla): new store listeners for devtools

### PR DESCRIPTION
**Pull request by dai-shi**
This supersedes #1965.

Two improvements:
- It doesn't notify on `restore`.
- It comes with `flushed` set.

Also, we added the notion of "revision" which means we will probably keep improving/changing in the future.

### CODE REVIEW

**Comment by dai-shi** [comment]
@arjunvegda check out the new one!

**Review by arjunvegda — COMMENTED** [review]
This is great! The write and restore functionality is working as expected. 

Is there a way to trigger `write-async` when a promise is resolved? Reference PR #1920

File: src/vanilla/store.ts [hunk 1]
```diff
@@ -518,7 +535,12 @@ export const createStore = () => {
         r = writeAtomState(a as AnyWritableAtom, ...args) as R
       }
       if (!isSync) {
-        flushPending()
+        const flushed = flushPending()
+        if (import.meta.env?.MODE !== 'production') {
+          storeListenersRev2.forEach((l) =>
+            l({ type: 'asyncWrite', flushed: flushed as Set<AnyAtom> })
```
[the remark this answers is not in the document]
Comment (dai-shi) [reply src/vanilla/store.ts:541]: Yes, this is about if `write` is sync or async. It has nothing to do with atom values.

**Comment by dai-shi** [comment]
> Is there a way to trigger `write-async` when a promise is resolved? Reference PR #1920

Our change here is about store listeners. Promise resolution isn't in it. (Our mental model is, Jotai store doesn't care about promise values.)
Basically, I still think you can handle promise resolution on your end.
We could add some kind of "promise listeners", but it just offload the task from `jotai-devtools` to `jotai`. (I mean, I don't think it adds any new capability. I might be missing something though. Maybe, if we want to track `continuePromise`. 🤔 )

**Comment by arjunvegda** [comment]
> Our mental model is, Jotai store doesn't care about promise values.

I get your perspective now! I considered promise values as a first-class citizen for Jotai, similar to any other non-promise value. So to me, it's natural to fire an event (`state` change for example) when the async value in the store gets updated, similar to how we do it for any other non-promise value. 😅 

`jotai-devtools` sounds like a better place for it then! 🙌

**Review by arjunvegda — APPROVED** [review]
LGTM!

**Comment by dai-shi** [comment]
Yeah, only the exception is that it wraps a promise to match with react behavior, like `continuePromise`, `resolvePromise`, `abortPromise` and so on. Now I think, ideally, this should go into `jotai/react` which would simplify `jotai` itself. We may try that road in the future, when we know more about react `use` best practice which doesn't exist yet.

**Comment by dai-shi** [comment]
@arjunvegda Can you try if/how `flushed` set helps in devtools?

**Comment by arjunvegda** [comment]
> @arjunvegda Can you try if/how `flushed` set helps in devtools?

This new API and access to `flushed` is a great addition. I played around with this API in devtools this weekend. I don't plan to use it immediately, perhaps after Time Travel is released. But here are some ideas 

- Create atom based subscription model for time travel to update the diff map instead of re-creating one every time there is a change
- Construct a map of atoms using these "atom level" subscriptions on the atom viewer tab and allow users to see diffs by atom 
- Maybe use this API to build a stack trace whenever there is a write event

---
This might be a stretch, but it'd be even nicer [in the future] to pre-hook for writing as well, maybe `pre-write` or `before-write`.  We could use those events to save stack traces on dev tools before the write occurs.

**Comment by dai-shi** [comment]
tbh, I thought about `pre-write` but I'm not a big fan of it.
I would like to offload as much as possible to `jotai-devtools` and make `jotai` as slim as possible.

So, an ideal way would wrap store.
```jsx
const store = devtools(createStore());

const App = () => {
  return (
    <Provider store={store}>
      <DevTools store={store} />
      {/* your app */}
    </Provider>
  );
};

// OR

const store = createStore();

const App = () => {
  return (
    <ProviderWithDevTools store={store}>
      {/* your app */}
    </ProviderWithDevTools>
  );
};
```
This doesn't work with the default store, so we would need a hook for it. 🤔 (Yeah, `setDefaultStore` may make sense.)

I wonder if we can eliminate `mountedAtom` in store.

**Review by arjunvegda — APPROVED** [review]
Thank you!!

**Pull request merged**